### PR TITLE
WIP: Holistic CLI test harness experiments

### PR DESCRIPTION
**Pull request by epage**
This includes
- Experiments with extension traits, opening the door for duct, see #32 
- API for quickly setting up a file system for testing, mostly inspired by my work on cobalt ("I'd like to test X but that'll be too much work to set up a fixture.  Maybe its good enough...").
- Comparisons with `cli_test_dir`, another holistic harness that is more rigid in its API

See commit messages for more details

Work to do
- [x] Replace `Assert`s ability to run commands with traits
- [x] Make it easy to setup a tempdir
- [ ] Add assertion predicates
- [ ] Determine crate(s) structure
- [ ] Tests and documentation

### CODE REVIEW

**Comment by azriel91** [comment]
btw, `tempdir` is now merged into [`tempfile`](https://crates.io/crates/tempfile) 3.0.0 and was published 2 days ago.

**Comment by killercup** [comment]
fyi @nikomatsakis just told me he used assert_cli and [manually added a tempdir wrapper](https://github.com/nikomatsakis/mathema/blob/8d5e006d9d1c436ed5b51035f90f400c62a58103/src/test/env.rs#L45-L82) because that was missing

File: src/cmd.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    /// Command::cargo_binary("assert_fixture")
+    ///     .output()
+    ///     .unwrap();
+    /// ```
+    fn cargo_binary<S: AsRef<ffi::OsStr>>(name: S) -> Self;
+}
+
+impl CommandCargoExt for process::Command {
+    fn main_binary() -> Self {
+        let mut cmd = process::Command::new("carg");
```
Comment (peddermaster2) [inline src/cmd.rs]: `carg` -> `cargo`?
Same in [line 51](https://github.com/assert-rs/assert_cli/pull/98/files#diff-518bd9db9fa60ba5991cb1df6489cb2fR51)

Comment (epage) [reply src/cmd.rs]: Thanks!

Obviously very little testing has been done which might be clear considering this is here :).  This is a rough sketch to get high level feedback before actually getting the details right.

**Comment by epage** [comment]
My current thought on crate organization
- `assert_cmd` for writing asserts on std::process::Command
- `assert_fs` for tempfile / path assertions
- `assert_cli` that re-exports both of the above

Couple questions from that
- I assume some people would appreciate `process::Command` helpers that this PR provides but aren't interested in assertions.  Should we advertise that we provide these or create yet another crate for them?
  - passing in a `Vec<u8>` for stdin
  - Turning an `Command` / `Output` into a `Result`
- Currently, the tempfile parts have a distinct struct for playing with files in the tempdir.  The cmd parts then support using that to run a command in that directory.  This might run afoul of trait rules if we split this into multiple crates.
  - Will it?
  - If it will, should we just switch to attaching these extension traits to `Path` instead?

File: src/cmd.rs [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+    /// use std::process::Command;
+    /// use assert_cli::cmd::*;
+    ///
+    /// Command::main_binary()
+    ///     .output()
+    ///     .unwrap();
+    /// ```
+    fn main_binary() -> Self;
+
+    /// Create a `Command` Run a specific binary of the current crate.
```
Comment (killercup) [inline src/cmd.rs]: s/Run/to run

File: src/cmd.rs [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+            .expect("Couldn't get mut ref to command stdin")
+            .write_all(&self.stdin)?;
+        Ok(spawned)
+    }
+}
+
+/// `std::process::Output` represented as a `Result`.
+pub type OutputResult = Result<process::Output, OutputError>;
+
+/// Extends `std::process::Output` with methods to to convert it to an `OutputResult`.
```
Comment (killercup) [inline src/cmd.rs]: s/to to/to

File: src/cmd.rs [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+    /// *(mirrors `std::process::Command::output`**
+    pub fn output(&mut self) -> io::Result<process::Output> {
+        self.spawn()?.wait_with_output()
+    }
+
+    /// Executes the command as a child process, returning a handle to it.
+    ///
+    /// By default, stdin, stdout and stderr are inherited from the parent.
+    ///
+    /// *(mirrors `std::process::Command::spawn`**
```
Comment (killercup) [inline src/cmd.rs]: s/**/)*

File: src/temp.rs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+    /// Access the path.
+    pub fn path(&self) -> &path::Path {
+        &self.path
+    }
+}
+
+/// Extend `TempDir` to run commands in it.
+pub trait TempDirCommandExt {
+    /// Constructs a new Command for launching the program at path program, with the following
```
Comment (killercup) [inline src/temp.rs]: s/path program/path

File: src/temp.rs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    ///
+    /// ```rust,ignore
+    /// extern crate assert_cli;
+    /// use assert_cli::temp::*;
+    ///
+    /// let temp = TempDir::new("TempDirChildExt_demo").unwrap();
+    /// temp.command("pwd").output().unwrap();
+    /// temp.close().unwrap();
+    /// ```
+    fn command<S>(&self, program: S) -> process::Command
```
Comment (killercup) [inline src/temp.rs]: i'd probably call this exec?

**Comment by epage** [comment]
This is now implemented as
- https://github.com/assert-rs/assert_fs
- https://github.com/assert-rs/assert_cmd

**Pull request closed without merging**